morango/sync/context.py

```diff
@@ -0,0 +1,224 @@
+        if transfer_session:
+            self.sync_session = transfer_session.sync_session
+            self.filter = transfer_session.get_filter() or self.filter
+            self.is_push = transfer_session.push
+
+    @property
+    def is_pull(self):
+        """
+        :rtype: bool
+        """
+        return not self.is_push
+
+    def __getstate__(self):
+        """Return dict of simplified data for serialization"""
+        return dict(
+            sync_session_id=self.sync_session.id if self.sync_session else None,
+            transfer_session_id=(
+                self.transfer_session.id if self.transfer_session else None
+            ),
+            filter=str(self.filter),
+            is_push=self.is_push,
+            stage=self.stage,
+            stage_status=self.stage_status,
+            capabilities=self.capabilities,
+        )
+
+    def __setstate__(self, state):
+        """Re-apply dict state after serialization"""
+        sync_session_id = state.get("sync_session_id", None)
+        if sync_session_id is not None:
+            self.sync_session = SyncSession.objects.get(pk=sync_session_id)
+
+        transfer_session_id = state.get("transfer_session_id", None)
+        if transfer_session_id is not None:
+            self.transfer_session = TransferSession.objects.get(pk=transfer_session_id)
+            if self.sync_session is None:
+                self.sync_session = self.transfer_session.sync_session
+
+        sync_filter = state.get("filter", None)
+        if sync_filter is not None:
+            self.filter = Filter(sync_filter)
+
+        self.is_push = state.get("is_push", None)
+        self.stage = state.get("stage", None)
+        self.stage_status = state.get("stage_status", None)
+        self.capabilities = state.get("capabilities", None)
+
+
+class LocalSessionContext(SessionContext):
+    """
+    Class that holds the context for operating on a transfer locally
+    """
+
+    __slots__ = (
+        "request",
+        "is_server",
+    )
+
+    def __init__(self, request=None, **kwargs):
+        """
+        :param request: If acting as the server, it should pass in the request, but the request
+            is not serialized into context. See `is_server` prop for determining if request was
+            passed in.
+        :type request: django.http.request.HttpRequest
+        """
+        capabilities = kwargs.pop("capabilities", [])
+        if request is not None:
+            capabilities = parse_capabilities_from_server_request(request)
+
+        super(LocalSessionContext, self).__init__(capabilities=capabilities, **kwargs)
+        self.request = request
+        self.is_server = request is not None
+
+    def update(self, **kwargs):
+        super(LocalSessionContext, self).update(**kwargs)
+        # this is a "local" session context, so the stage and status should be in sync with the
+        # local transfer session object
```

Comment:
If they should always be in sync, could potentially implement them as properties?

Reply:
(but I don't have my head fully wrapped around the flow, so what I'm saying may be off)

Reply:
Yeah I think making them properties makes a lot of sense.